Feature/lighting

martinfalk opened the pull request:
Changes proposed in this PR:
 * reworked `PositionProperty` to have a consistent internal state
 * `PositionProperty` now features read-only properties representing the position in various coordinate spaces
 * added read-only argument to OrdinalRefProperty and OrdinalRefPropertyState
 * minor refactoring of the camera widget (visibility of the widget and buttons for moving the camera)
![image](https://github.com/inviwo/inviwo/assets/9251300/a8b0dac1-be2d-47db-8bfd-b16796ddac7a)
--- 

Updated `PositionProperty` with different outputs
![image](https://github.com/inviwo/inviwo/assets/9251300/bb355076-871f-4819-9843-00babe206f54)

martinfalk committed:
UserInterfaceGL: minor refactoring of camera widget

martinfalk committed:
Core,Base: Trackball minor refactoring

martinfalk committed:
Core: made OrdinalRefProperty more consistent

martinfalk committed:
Core,Base: refactored PositionProperty

* expose resulting position in different coordinate spaces

martinfalk committed:
BaseGL: updated lighting raycaster workspaces

martinfalk committed:
Core,Base: clang-tidy

ResearchDaniel commented on include/inviwo/core/properties/ordinalrefproperty.h:
```diff
@@ -274,6 +281,45 @@ namespace util {
+                ConstraintBehavior::Ignore,
+                T{minMax},
+                ConstraintBehavior::Ignore,
+                T{static_cast<V>(0.1)},
```
```suggestion
                T{static_cast<V>(0.001 * static_cast<double>(minMax))},
```

petersteneteg replied:
I think this should be consistent with the ordinal ones.
I suspect that the code suggested will create warnings...

martinfalk replied:
This is in line with the implementation in  [`ordinalproperty.h`](https://github.com/inviwo/inviwo/blob/master/include/inviwo/core/properties/ordinalproperty.h)

https://github.com/inviwo/inviwo/blob/96f60e567482831f60dd481ffc294f892c01e840/include/inviwo/core/properties/ordinalproperty.h#L296-L307

ResearchDaniel commented on include/inviwo/core/properties/ordinalrefproperty.h:
```diff
@@ -274,6 +281,45 @@ namespace util {
+                ConstraintBehavior::Ignore,
+                T{minMax},
+                ConstraintBehavior::Ignore,
+                T{static_cast<V>(1)},
```
Is it intentional to have 1 here and 0.1 above?
```suggestion
                T{static_cast<V>(0.01 * minMax)},
```

ResearchDaniel commented on include/inviwo/core/properties/ordinalrefproperty.h:
```diff
@@ -274,6 +281,45 @@ namespace util {
+                ConstraintBehavior::Ignore,
+                T{minMax},
+                ConstraintBehavior::Ignore,
+                T{static_cast<V>(1)},
```
```suggestion
                T{static_cast<V>(0.01 * minMax)},
```

martinfalk replied:
yes, this is the signed int case. The first part deals with floating point types.

ResearchDaniel commented on include/inviwo/core/properties/positionproperty.h:
```diff
@@ -53,54 +58,83 @@ class IVW_CORE_API PositionProperty : public CompositeProperty {
+     * Set the property semantics of the position property to \p semantics
      */
-    void set(const vec3& worldSpacePos);
+    void setPositionSemantics(PropertySemantics semantics);
```
Override or use setSemantics

martinfalk replied:
But I don't want to mess with the overall semantics of the composite property!

ResearchDaniel replied:
I see. It is a bit double meaning with the existence of a separate PositionProperty

ResearchDaniel commented on modules/base/src/processors/directionallightsourceprocessor.cpp:
```diff
@@ -113,19 +111,9 @@ void DirectionalLightSourceProcessor::process() {
-            break;
-    }
+    const vec3 lightPos = lightPosition_.get(CoordinateSpace::World);
+    const vec3 dir = -glm::normalize(lightPosition_.getWorldSpaceDirection());
```
Nice that the duplicate code is removed, but it is not very clear that the camera look at position  is used here

martinfalk replied:
Should be resolved now. The direction is now derived from the current position and offset (origin, camera, or custom).

ResearchDaniel commented on modules/base/src/processors/pointlightsourceprocessor.cpp:
```diff
@@ -150,31 +147,17 @@ PointLightSourceProcessor::PointLightSourceProcessor()
-        }
-    }
+    const vec3 lightPos = lightPosition_.get(CoordinateSpace::World);
+    const vec3 dir = -glm::normalize(lightPosition_.getWorldSpaceDirection());
```
Same as other

ResearchDaniel commented on modules/base/src/processors/pointlightsourceprocessor.cpp:
```diff
@@ -316,14 +297,17 @@ void PointLightInteractionHandler::setLightPosFromScreenCoords(const vec2& norma
     auto res = raySphereIntersection(vec3(0.f), sceneRadius, rayOrigin, rayDir, 0.0f,
```
Having origin here might be a good assumption. Better to use lookTo?

ResearchDaniel commented on modules/base/src/processors/spotlightsourceprocessor.cpp:
```diff
@@ -106,21 +103,9 @@ void SpotLightSourceProcessor::process() {
-        }
-    }
+    const vec3 lightPos = lightPosition_.get(CoordinateSpace::World);
+    const vec3 dir = -glm::normalize(lightPosition_.getWorldSpaceDirection());
```
Same here

ResearchDaniel commented on resources/svgicons/camera-dolly-closer.svg:
camera icon plus direction would probably be easier to understand. Or even more explicit:
https://www.iconfinder.com/icons/3625387/camera_equipment_film_movie_slider_dolly_tool_video_icon

martinfalk replied:
But it's neither magnification nor minification (plus/minus). Dolly refers to moving the camera closer/away.

ResearchDaniel commented on resources/svgicons/masters/camera-movement.svg:
I don’t understand this icon. Is it more something like
https://www.flaticon.com/free-icon/move_271222?term=move&page=1&position=1&origin=search&related_id=271222

martinfalk replied:
This is the master icon for the other ones. Therefore it contains multiple layers, one for each direction plus dolly.

ResearchDaniel commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,222 @@
+                      {{"world", "World coordinates", CoordinateSpace::World},
+                       {"view", "View coordinates", CoordinateSpace::View},
+                       {"clip", "Clip coordinates", CoordinateSpace::Clip}})
+    , transformPosition_{"transformPosition", "Transform Position to Reference Space",
```
Is there a use case when this is not desired?

petersteneteg replied:
I would also argue for having this aways on and skip the setting...
if you can't to you can always copy the position then change space and paste. 
But I think that is very rare.

martinfalk replied:
Removed the property.

ResearchDaniel commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,222 @@
+vec3 PositionProperty::getWorldSpaceDirection() const {
+    if (camera_ &&
+        (referenceSpace_ == CoordinateSpace::View || referenceSpace_ == CoordinateSpace::Clip)) {
+        return convertToWorld(position_, referenceSpace_) - camera_->getLookTo();
```
Should be reversed?

martinfalk replied:
No, it is the direction from the origin/look to toward the position.

ResearchDaniel replied:
Ok, if we want to have this function in the position property we should update the docs to something like:
```
/* Determines the direction from the origin to the position in the current coordinate space and
 * transforms it to world coordinates. If the property has an associated camera, the origin is the 
 * camera lookTo for view or clip space, otherwise (0, 0, 0).
 * In case of world coordinates, this is equivalent to the position itself.
 * @return position - origin in world space.
 */
```

petersteneteg replied:
if you return a direction it should probably be normalized? 
is worldspace the only needs space? for this?

ResearchDaniel replied:
I think that it would be better to have a new `DirectionPoperty` that has this functionality, but I also understand that it could be outside of the scope of this PR.

martinfalk replied:
> if you return a direction it should probably be normalized? is worldspace the only needed space for this?

At the moment yes, since all of our lighting calculations take place in world space and therefore light position and direction is to be specified in world coords.

martinfalk replied:
Change function signature to `vec3 getDirection(CoordinateSpace space) const;`. The direction corresponds to the normalized vector between the position and the current offset (either (0,0,0), camera look at, or custom).

ResearchDaniel commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,222 @@
+            return pos;
+        case CoordinateSpace::View:
+            if (camera_) {
+                return vec3{camera_->viewMatrix() * vec4{pos, 1.0f}};
```
I assume it is not possible to use SpatialCameraCoordinateTransformerImpl

petersteneteg replied:
SpatialCameraCoordinateTransformerImpl should work here

martinfalk replied:
Without a `SpatialEntity` at hand? 
https://github.com/inviwo/inviwo/blob/96f60e567482831f60dd481ffc294f892c01e840/include/inviwo/core/datastructures/coordinatetransformer.h#L537-L542

martinfalk replied:
Added a identity SpatialEntity to be able to use the transformer.

petersteneteg commented on include/inviwo/core/interaction/trackballobject.h:
```diff
@@ -40,27 +40,28 @@ namespace inviwo {
-    virtual const vec3& getLookTo() const = 0;
-    virtual const vec3& getLookFrom() const = 0;
-    virtual const vec3& getLookUp() const = 0;
+    [[nodiscard]] virtual vec3 getLookTo() const = 0;
```
think we can disable the nodiscard warning. It seems excessive to add to every getter...

petersteneteg commented on include/inviwo/core/properties/cameraproperty.h:
```diff
@@ -95,11 +95,11 @@ class IVW_CORE_API CameraProperty : public CompositeProperty, public TrackballOb
     CameraProperty& setCamera(std::unique_ptr<Camera> camera);
 
-    virtual const vec3& getLookFrom() const override;
+    virtual vec3 getLookFrom() const override;
```
Why?

martinfalk replied:
The point light source processor has a trackball implementation and needs to return a look from position. This position is coming from the PositionProperty and are computed on-the-fly. Hence, no reference.

petersteneteg commented on include/inviwo/core/properties/positionproperty.h:
```diff
@@ -53,54 +58,83 @@ class IVW_CORE_API PositionProperty : public CompositeProperty {
+                     CoordinateSpace coordinateSpace = CoordinateSpace::World,
                      CameraProperty* camera = nullptr,
                      InvalidationLevel = InvalidationLevel::InvalidResources,
                      PropertySemantics semantics = PropertySemantics::Default);
```
This should be PropertySemantics::Default nothing else. 
This means that you get the default widget for a PositionProperty. that should be the default option. 
If you want something else change what the default means for a PositionProperty

petersteneteg commented on include/inviwo/core/properties/positionproperty.h:
```diff
@@ -53,54 +58,83 @@ class IVW_CORE_API PositionProperty : public CompositeProperty {
-    vec3 positionWorldSpace_;  //< Used for always keeping track of the current position in world
-                               // space.
-    CameraProperty* camera_;   //< Non-owning reference.
+    std::vector<Property*> outputProperties_;
```
remove, use a std::tie props thing, or a function that returns an array of the 4 pointers...

petersteneteg commented on modules/base/src/processors/diffuselightsourceprocessor.cpp:
```diff
@@ -81,9 +81,7 @@ DiffuseLightSourceProcessor::DiffuseLightSourceProcessor()
-                     FloatVec3Property("position", "Position", vec3(1.f, 0.65f, 0.65f), vec3(-10.f),
-                                       vec3(10.f)),
-                     &camera_)
+                     vec3(100.0f, 65.0f, 65.0f), CoordinateSpace::World, &camera_)
```
set semantics here too....

petersteneteg commented on modules/base/src/processors/directionallightsourceprocessor.cpp:
```diff
@@ -82,10 +82,8 @@ DirectionalLightSourceProcessor::DirectionalLightSourceProcessor()
-          "lightPosition", "Light Source Position", "Origin of the light source"_help,
-          FloatVec3Property("position", "Position", vec3(100.f), vec3(-100.f), vec3(100.f)),
-          &camera_)
+    , lightPosition_("lightPosition", "Light Source Position", "Origin of the light source"_help,
```
set semantics here too....

petersteneteg commented on modules/base/src/processors/pointlightsourceprocessor.cpp:
```diff
@@ -95,10 +95,8 @@ PointLightSourceProcessor::PointLightSourceProcessor()
-                     FloatVec3Property("position", "Position", vec3(-2.f, -50.f, 90.f),
-                                       vec3(-100.f), vec3(100.f)),
-                     &camera_)
+    , lightPosition_("lightPosition", "Light Source Position", vec3(-2.f, -50.f, 90.f),
```
set semantics here too....

petersteneteg commented on modules/base/src/processors/spotlightsourceprocessor.cpp:
```diff
@@ -73,10 +73,8 @@ SpotLightSourceProcessor::SpotLightSourceProcessor()
-          "lightPosition", "Light Source Position",
-          FloatVec3Property("position", "Position", vec3(100.f), vec3(-100.f), vec3(100.f)),
-          &camera_)
+    , lightPosition_("lightPosition", "Light Source Position", vec3(100.f), CoordinateSpace::World,
```
set semantics here too....

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,222 @@
+void PositionProperty::invalidateOutputProperties() {
+    for (auto p : outputProperties_) {
+        p->propertyModified();
+    }
```
```suggestion
    for (auto& p : {worldPos_, viewPos_, clipPos_, screenPos_}) {
        p.propertyModified();
    }
```

petersteneteg commented:
disable 
readability-uppercase-literal-suffix
modernize-use-nodiscard 
in .clang-tidy

martinfalk committed:
Core: identity implementation for SpatialEntity

martinfalk committed:
Core: requested changes

martinfalk committed:
QtWidgets: light position widget fix

martinfalk committed:
Base: orientation indicator fix

* muted default colors

martinfalk committed:
Core: PositionProperty with custom offset

martinfalk committed:
Base: requested changes

martinfalk committed:
QtWidgets: spherical coords fix

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,260 @@
-    // Add properties
-    addProperty(referenceFrame_);
-    addProperty(position_);
+    referenceSpace_.setSelectedValue(coordinateSpace);
```
this could be done in the constructor using a lambda and a optionpropertystate

petersteneteg replied:
also remove it here...

petersteneteg replied:
still here?

ResearchDaniel replied:
Unresolved this

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,260 @@
+                return vec3{transformer.getClipToWorldMatrix() * vec4{pos, 1.0f}};
+            } else {
+                return pos;
+            }
```
Would something like this not work? 
```suggestion
vec3 PositionProperty::convertToWorld(const vec3& pos, CoordinateSpace sourceSpace) const {
     if (!camera_) {
          return pos;
     }
     SpatialIdentity identity;
     SpatialCameraCoordinateTransformerImpl transformer{identity, camera_->get()};
     return vec3{transformer.getMatrix(sourceSpace, CoordinateSpace::World) * vec4{pos, 1.0f}};
}   
```

We might add a vec3 transformPosition(from, to, vec3 pos) to SpatialCoordinateTransformer 
and maybe a 
vec3 transformDirection(from, to, vec3 pos)

martinfalk replied:
good suggestion

martinfalk committed:
Qt,QtWidgets,AnimationQt,OpenGLQt,Python3Qt: named toolbars

martinfalk committed:
Core,Base: light direction refactoring

martinfalk committed:
Core: light soure info

* camelcased LightSourceType enum

martinfalk committed:
Core: PositionProperty fix

martinfalk committed:
Core: SimpleLightingProperty uses PositionProperty

martinfalk committed:
BaseGL: heightfield renderer fix

martinfalk committed:
Testing,Examples: update workspaces and reference images

martinfalk committed:
Meta: Changelog update

petersteneteg commented on data/workspaces/boron.inv:
```diff
@@ -154,21 +152,32 @@
+                                    </Property>
+                                    <Property type="org.inviwo.CompositeProperty" identifier="output">
+                                        <Properties />
+                                        <collapsed content="1" />
```
Default state?

petersteneteg commented on data/workspaces/boron.inv:
```diff
@@ -154,21 +152,32 @@
+                                <Properties>
+                                    <Property type="org.inviwo.OptionPropertyEnumInt" identifier="referenceFrame">
+                                        <selectedIdentifier content="view" />
+                                    </Property>
```
Deafult?

martinfalk replied:
Default state is <s>world space</s> view space.

petersteneteg commented on data/workspaces/boron.inv:
```diff
@@ -264,6 +268,11 @@
                         <increment content="1" />
                         <value content="0" />
                     </Property>
+                    <Property type="org.inviwo.DirectoryProperty" identifier="layerDir">
```
this should probably be considered a default state https://github.com/inviwo/inviwo/blob/96f60e567482831f60dd481ffc294f892c01e840/src/core/properties/fileproperty.cpp#L381 :/

petersteneteg replied:
this is my fault I think...

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,240 @@
+                                      {"view", "View coordinates", CoordinateSpace::View},
+                                      {"clip", "Clip coordinates", CoordinateSpace::Clip}}}
+                          .set("Reference coordinate space for the position below."_help)
+                          .set([coordinateSpace]() -> size_t {
```
Might add a OptionPropertyState::set(const T& newSelectedValue) method to make this a bit easier....

martinfalk replied:
Might be tricky for an `OptionPropertInt`, as `set` will clash with `set(size_t index)`

https://github.com/inviwo/inviwo/blob/96f60e567482831f60dd481ffc294f892c01e840/include/inviwo/core/properties/optionproperty.h#L128-L131

petersteneteg replied:
Might name it setSelectedValue?

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,240 @@
+    vec4 result{transformer.getMatrix(CoordinateSpace::World, targetSpace) * vec4{pos, 1.0f}};
+    if (targetSpace == CoordinateSpace::Clip) {
+        result /= result.w;
+    }
```
We should really have this in the transformer...

petersteneteg commented on modules/basegl/include/modules/basegl/processors/lightvolumeraycaster.h:
```diff
@@ -0,0 +1,69 @@
+    IsoTFComponent<1> isoTF_;
+    RaycastingComponent raycasting_;
+    CameraComponent camera_;
+    LightComponent light_;
```
Why both a lightComponent and a LightVolumeComponent?

martinfalk replied:
LightComponent provides only the SimpleLightProperty settings for regular illumination via APPLY_LIGHTING. LightVolumeComponent is independent of this one and adds a light volume inport plus modifies the color _after_ illumination, before compositing.

petersteneteg replied:
Ok I was just wondering it it would not be easier to integrate if you replace the LightComponent with a new LightVolumeComponent that has all of that? Should there even be a SimpleLightProperty in there? should that not come from the light volume?

martinfalk replied:
Moved to separate PR.

ResearchDaniel commented:
@martinfalk Can the approved changes be merged first before adding the new lighting components?

martinfalk commented:
> @martinfalk Can the approved changes be merged first before adding the new lighting components?

Guess so.

martinfalk committed:
Core: extended OptionPropertyState

* setter for setting the index based on value of type T

martinfalk committed:
Core: added transforms to SpatialCoordinateTransformer

* added transform functions for positions and normals

martinfalk committed:
BaseGL: LightVolumeGL directional light fix

martinfalk committed:
Core: utility function for transforming vec3

petersteneteg commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,226 @@
+                                      {"view", "View coordinates", CoordinateSpace::View},
+                                      {"clip", "Clip coordinates", CoordinateSpace::Clip}}}
+                          .set("Reference coordinate space for the position below."_help)
+                          .setSelectedValue(coordinateSpace)}
```
Nice!

ResearchDaniel commented on src/core/properties/positionproperty.cpp:
```diff
@@ -28,115 +28,226 @@
+                util::ordinalSymmetricVector(position, 10.0f)
+                    .set("Spatial position in the reference coordinate space"_help)
+                    .set(positionSemantics)}
+    , coordinateOffsetMode_("coordinateOffsetMode", "Offset",
```
Nice that this is explicit!

ResearchDaniel reviewed (changes requested):
Unresolved a comment that was not addressed. A code line should be removed

martinfalk committed:
Core: include reference space in LightingConfig

martinfalk commented:
Recent changes
* Added utility position and normal transform functions to util/glm.h in [cc38577](https://github.com/inviwo/inviwo/pull/1512/commits/cc385771ef79c758c38f4851e1f3590726429a81)
* fixed issues with directional light and Light Volume processor
* default light position set to View and located to the top-right of the camera
* updated regression tests and example workspaces

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg reviewed (dismissed):
Next time, put any regression test workspace changes in a different PR, now there is know way to know if a image is because you changes the workspace, or because of code changes, or anything else.

martinfalk committed:
Core: use default shading mode in SimpleLightingProperty

martinfalk committed:
Testing: updated reference images

* difference due Windows/MacOS
